File: src/stats.py
import os
import shutil
import ssl
import subprocess
import urllib.request
from typing import Tuple


def get_service_status(service_name: str) -> str:
    try:
        res = subprocess.run(
            ["systemctl", "is-active", service_name],
            capture_output=True,
            text=True,
            timeout=2
        )
        status = res.stdout.strip()
        if status == "active":
            return "active (работает)"
        return f"inactive ({status})" if status else "не запущен"
    except Exception:
        return "неизвестно"


def get_bbr_status() -> str:
    try:
        with open("/proc/sys/net/ipv4/tcp_congestion_control", "r") as f:
            cc = f.read().strip()
        if "bbr" in cc:
            return "включен (bbr)"
        return f"выключен ({cc})"
    except Exception:
        return "неизвестно"


def check_wast_status(domain: str, sub_dir: str) -> Tuple[str, str]:
    stub_msg = "нет ответа"
    try:
        ctx = ssl._create_unverified_context()
        req = urllib.request.Request("https://127.0.0.1:8443/", headers={"Host": domain})
        with urllib.request.urlopen(req, context=ctx, timeout=2) as resp:
            if resp.status == 200:
                stub_msg = "доступна (200 OK)"
            else:
                stub_msg = f"код {resp.status}"
    except Exception:
        stub_msg = "ошибка подключения"

    sub_count = 0
    if os.path.isdir(sub_dir):
        sub_count = len([f for f in os.listdir(sub_dir) if f.endswith(".txt")])

    return stub_msg, f"активны ({sub_count} шт.)"


def get_cert_expiry(domain: str) -> str:
    cert_path = f"/etc/letsencrypt/live/{domain}/fullchain.pem"
    if not os.path.isfile(cert_path):
        return "отсутствует"
    try:
        res = subprocess.run(
            ["openssl", "x509", "-enddate", "-noout", "-in", cert_path],
            capture_output=True,
            text=True,
            timeout=2
        )
        out = res.stdout.strip()
        if "notAfter=" in out:
            return out.split("notAfter=", 1)[1].strip()
        return out
    except Exception:
        return "ошибка проверки"


def get_uptime() -> str:
    try:
        with open("/proc/uptime", "r") as f:
            sec = float(f.readline().split()[0])
        days = int(sec // 86400)
        hours = int((sec % 86400) // 3600)
        minutes = int((sec % 3600) // 60)
        parts = []
        if days > 0:
            parts.append(f"{days} дн.")
        if hours > 0:
            parts.append(f"{hours} ч.")
        parts.append(f"{minutes} мин.")
        return " ".join(parts)
    except Exception:
        return "недоступно"


def get_cpu_info() -> str:
    try:
        load1, load5, load15 = os.getloadavg()
        cores = os.cpu_count() or 1
        return f"{load1:.2f}, {load5:.2f}, {load15:.2f} (ядер: {cores})"
    except Exception:
        return "недоступно"


def get_ram_info() -> str:
    try:
        mem = {}
        with open("/proc/meminfo", "r") as f:
            for line in f:
                parts = line.split(":")
                if len(parts) == 2:
                    mem[parts[0].strip()] = int(parts[1].split()[0])
        total = mem.get("MemTotal", 0)
        avail = mem.get("MemAvailable", 0)
        used = total - avail
        pct = (used / total * 100) if total else 0
        return f"{used / 1024:.0f} MB / {total / 1024:.0f} MB ({pct:.1f}%)"
    except Exception:
        return "недоступно"


def get_disk_info() -> str:
    try:
        total, used, free = shutil.disk_usage("/")
        pct = (used / total * 100) if total else 0
        return f"{used / (1024**3):.1f} GB / {total / (1024**3):.1f} GB ({pct:.1f}%)"
    except Exception:
        return "недоступно"

File: src/stats.go
package main

import (
	"context"
	"crypto/tls"
	"crypto/x509"
	"encoding/pem"
	"fmt"
	"net/http"
	"os"
	"os/exec"
	"path/filepath"
	"runtime"
	"strconv"
	"strings"
	"syscall"
	"time"
)

func GetServiceStatus(serviceName string) string {
	ctx, cancel := context.WithTimeout(context.Background(), 2*time.Second)
	defer cancel()

	cmd := exec.CommandContext(ctx, "systemctl", "is-active", serviceName)
	out, err := cmd.Output()
	status := strings.TrimSpace(string(out))
	if err != nil {
		if status != "" {
			return fmt.Sprintf("inactive (%s)", status)
		}
		return "не запущен"
	}

	if status == "active" {
		return "active (работает)"
	}
	if status != "" {
		return fmt.Sprintf("inactive (%s)", status)
	}
	return "не запущен"
}

func GetBBRStatus() string {
	data, err := os.ReadFile("/proc/sys/net/ipv4/tcp_congestion_control")
	if err != nil {
		return "неизвестно"
	}
	cc := strings.TrimSpace(string(data))
	if strings.Contains(cc, "bbr") {
		return "включен (bbr)"
	}
	return fmt.Sprintf("выключен (%s)", cc)
}

func CheckWastStatus(domain, subDir string) (string, string) {
	stubMsg := "нет ответа"
	tr := &http.Transport{
		TLSClientConfig: &tls.Config{
			InsecureSkipVerify: true,
			ServerName:         domain,
		},
	}
	client := &http.Client{
		Transport: tr,
		Timeout:   2 * time.Second,
	}

	req, err := http.NewRequest("GET", "https://127.0.0.1:8443/", nil)
	if err == nil {
		req.Host = domain
		resp, err := client.Do(req)
		if err == nil {
			if resp.StatusCode == 200 {
				stubMsg = "доступна (200 OK)"
			} else {
				stubMsg = fmt.Sprintf("код %d", resp.StatusCode)
			}
			resp.Body.Close()
		} else {
			stubMsg = "ошибка подключения"
		}
	}

	subCount := 0
	entries, err := os.ReadDir(subDir)
	if err == nil {
		for _, e := range entries {
			if !e.IsDir() && strings.HasSuffix(e.Name(), ".txt") {
				subCount++
			}
		}
	}

	return stubMsg, fmt.Sprintf("активны (%d шт.)", subCount)
}

func GetCertExpiry(domain string) string {
	certPath := filepath.Join("/etc/letsencrypt/live", domain, "fullchain.pem")
	data, err := os.ReadFile(certPath)
	if err != nil {
		return "отсутствует"
	}

	block, _ := pem.Decode(data)
	if block == nil {
		return "ошибка проверки"
	}

	cert, err := x509.ParseCertificate(block.Bytes)
	if err != nil {
		return "ошибка проверки"
	}

	return cert.NotAfter.Format("Jan _2 15:04:05 2006 MST")
}

func GetUptime() string {
	data, err := os.ReadFile("/proc/uptime")
	if err != nil {
		return "недоступно"
	}

	fields := strings.Fields(string(data))
	if len(fields) == 0 {
		return "недоступно"
	}

	sec, err := strconv.ParseFloat(fields[0], 64)
	if err != nil {
		return "недоступно"
	}

	totalSec := int64(sec)
	days := totalSec / 86400
	hours := (totalSec % 86400) / 3600
	minutes := (totalSec % 3600) / 60

	var parts []string
	if days > 0 {
		parts = append(parts, fmt.Sprintf("%d дн.", days))
	}
	if hours > 0 {
		parts = append(parts, fmt.Sprintf("%d ч.", hours))
	}
	parts = append(parts, fmt.Sprintf("%d мин.", minutes))

	return strings.Join(parts, " ")
}

func GetCPUInfo() string {
	data, err := os.ReadFile("/proc/loadavg")
	if err != nil {
		return "недоступно"
	}

	fields := strings.Fields(string(data))
	if len(fields) < 3 {
		return "недоступно"
	}

	cores := runtime.NumCPU()
	return fmt.Sprintf("%s, %s, %s (ядер: %d)", fields[0], fields[1], fields[2], cores)
}

func GetRAMInfo() string {
	data, err := os.ReadFile("/proc/meminfo")
	if err != nil {
		return "недоступно"
	}

	var totalKB, availKB float64
	lines := strings.Split(string(data), "\n")
	for _, line := range lines {
		parts := strings.Split(line, ":")
		if len(parts) != 2 {
			continue
		}
		key := strings.TrimSpace(parts[0])
		valParts := strings.Fields(strings.TrimSpace(parts[1]))
		if len(valParts) == 0 {
			continue
		}
		v, _ := strconv.ParseFloat(valParts[0], 64)
		if key == "MemTotal" {
			totalKB = v
		} else if key == "MemAvailable" {
			availKB = v
		}
	}

	if totalKB == 0 {
		return "недоступно"
	}

	usedKB := totalKB - availKB
	pct := (usedKB / totalKB) * 100
	return fmt.Sprintf("%.0f MB / %.0f MB (%.1f%%)", usedKB/1024, totalKB/1024, pct)
}

func GetDiskInfo() string {
	var stat syscall.Statfs_t
	if err := syscall.Statfs("/", &stat); err != nil {
		return "недоступно"
	}

	totalBytes := stat.Blocks * uint64(stat.Bsize)
	freeBytes := stat.Bfree * uint64(stat.Bsize)
	usedBytes := totalBytes - freeBytes

	totalGB := float64(totalBytes) / float64(1024*1024*1024)
	usedGB := float64(usedBytes) / float64(1024*1024*1024)
	pct := 0.0
	if totalBytes > 0 {
		pct = (float64(usedBytes) / float64(totalBytes)) * 100
	}

	return fmt.Sprintf("%.1f GB / %.1f GB (%.1f%%)", usedGB, totalGB, pct)
}
